etna/models/nn/tft_native/layers.py

```diff
@@ -0,0 +1,293 @@
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward pass.
+
+        Parameters
+        ----------
+        x:
+            batch of data with shapes (batch_size, num_timestamps, input_size)
+
+        Returns
+        -------
+        :
+            output batch of data with shapes (batch_size, num_timestamps, output_size)
+        """
+        x = self.dropout(x)
+        a = self.activation_fc(x)
+        b = self.sigmoid(self.gated_fc(x))
+        x = torch.mul(a, b)
+        return x
+
+
+class GateAddNorm(nn.Module):
+    """Gated Add&Norm layer."""
+
+    def __init__(self, input_size: int, output_size: int, dropout: float = 0.1) -> None:
+        """Init Add&Norm layer.
+
+        Parameters
+        ----------
+        input_size:
+            input size of the feature representation
+        output_size:
+            output size of the feature representation
+        dropout:
+            dropout rate
+        """
+        super().__init__()
+
+        self.input_size = input_size
+        self.output_size = output_size
+        self.dropout = dropout
+
+        self.glu = GatedLinearUnit(input_size=self.input_size, output_size=self.output_size, dropout=self.dropout)
+        self.norm = nn.LayerNorm(self.output_size)
+
+    def forward(self, x: torch.Tensor, residual: torch.Tensor) -> torch.Tensor:
+        """Forward pass.
+
+        Parameters
+        ----------
+        x:
+            batch of data with shapes (batch_size, num_timestamps, input_size)
+        residual:
+            batch of data passed through skip connection with shapes (batch_size, num_timestamps, output_size)
+        Returns
+        -------
+        :
+            output batch of data with shapes (batch_size, num_timestamps, output_size)
+        """
+        x = self.glu(x)
+        x = self.norm(x + residual)
+        return x
+
+
+class GatedResidualNetwork(nn.Module):
+    """Gated Residual Network (GRN)."""
+
+    def __init__(self, input_size: int, output_size: int, dropout: float = 0.1, context: bool = False) -> None:
+        """Init GRN.
+
+        Parameters
+        ----------
+        input_size:
+            input size of the feature representation
+        output_size:
+            output size of the feature representation
+        dropout:
+            dropout rate
+        context:
+            whether to pass context vector through the block
+        """
+        super().__init__()
+        self.input_size = input_size
+        self.output_size = output_size
+        self.dropout = dropout
+        self.context = context
+
+        self.fc1 = nn.Linear(self.input_size, self.input_size)
+        self.elu = nn.ELU()
+
+        if self.context:
+            self.context_fc = nn.Linear(self.input_size, self.input_size, bias=False)
+
+        self.residual_fc = nn.Linear(self.input_size, self.output_size) if self.input_size != self.output_size else None
```

Comment:
What is it?
In paper notation all the linear layers shouldn't change the shape, the same in GLU input_size=output_size
<img width="229" alt="Снимок экрана 2024-02-09 в 09 04 27" src="https://github.com/etna-team/etna/assets/55380696/04730467-ebf5-4e6f-95e1-c71bab7c2052">

Reply:
There are also words about "weights sharing"
<img width="397" alt="Снимок экрана 2024-02-09 в 09 06 47" src="https://github.com/etna-team/etna/assets/55380696/8d7b0b8a-880f-4ade-abb4-77b78c09bee4">

Doesn't it imply that all the layers should share weights?

Reply:
UPD: it means weights sharing across time dim

Reply:
<img width="300" alt="image" src="https://github.com/etna-team/etna/assets/82458209/15936a54-aa72-4712-bf4a-643c7f741048">

We need reshape of residual data in Variable Selection Block, because in the beginning it will be `num_features * hidden_size`, but in the end it must be `num_features`